ocaml/runtime/fiber.c

```diff
@@ -111,7 +114,34 @@ Caml_inline struct stack_info* alloc_for_stack (mlsize_t wosize)
   si->size = len;
   return si;
 #else
+#ifdef NATIVE_CODE
+  size_t bsize = Bsize_wsize(wosize);
+  int page_size = getpagesize();
+  int num_pages = (bsize + page_size - 1) / page_size;
+  bsize = (num_pages + 2) * page_size;
+  size_t len = sizeof(struct stack_info) +
+               bsize +
+               8 /* for alignment to 16-bytes, needed for arm64 */ +
+               sizeof(struct stack_handler);
+  struct stack_info* block;
+  int memalign_res = posix_memalign((void **) &block, page_size, len);
+  if (memalign_res) {
+    return NULL;
+  }
+  block->size = len;
+  int protect = mprotect(((char*) block) + page_size, page_size, PROT_NONE);
```

Comment:
It's probably fine, but there's definitely something a bit weird about mprotecting memory that you got from the allocator. Have you considered using mmap directly instead? For instance, by extending the USE_MMAP_MAP_STACK case above with a guard page?

Reply:
Mostly, it struck me as impolite!

Concretely, it'll kill transparent hugepages if the allocator gives you a region in the middle of a large allocated area that had THP optimisation applied by the kernel.